### PR TITLE
ParticleContainer::make_alike

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

This creates a new AMReX particle container type with same compile-time and run-time attributes. But, it can change its allocator. This is helpful when creating temporary particle buffers for filter operations and device-to-host copies.

## Additional background

https://github.com/ECP-WarpX/WarpX/pull/2860

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [x] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by atmyers** [comment]
What do you think about making this a static member function, so it would be used like

```
auto new_pc = ParticleContainer::make_alike<amrex::PinnedArenaAllocator>(pc);
```

i.e. more similar to, `std::make_unique`, etc...

**Comment by ax3l** [comment]
Yes, I thought about this as well but that implementation needed a `const_cast` if `pc` is passed as a `const` argument. That's indeed how I first added it to WarpX:
https://github.com/ECP-WarpX/WarpX/pull/2860/commits/71d33f41f37f01f07735c75b7084dd081c952695

```C++
    static ParticleContainer
    make_alike (ParticleContainer const * pc)
    {
        auto gdb = const_cast<amrex::ParGDBBase*>(pc->GetParGDB());
        ParticleContainer tmp(gdb);

        // add runtime real comps to tmp
        for (int ic = 0; ic < pc->NumRuntimeRealComps(); ++ic) { tmp.AddRealComp(false); }

        // add runtime int comps to tmp
        for (int ic = 0; ic < pc->NumRuntimeIntComps(); ++ic) { tmp.AddIntComp(false); }

        return tmp;
    }
```
but let me try again here, the original code was more templated.

I think the main issue resides that we would like to pass `pc` as `ParticleContainer const *` not `ParticleContainer *`.

Another nitch: since functions do not support partial specialization, this needs to read somthing like:
```C++
auto tmp = ParticleContainer<..., amrex::PinnedArenaAllocator>::make_alike(pc);
```

Alternatively, we could maybe do an interface of the form:
```C++
auto tmp = ParticleContainer::make_alike(pc, amrex::PinnedArenaAllocator{});
```
...but the `const_cast` need would remain for the `GetParGDB()`.

**Comment by atmyers** [comment]
As discussed offline, let's just leave this as a non-static member function, since it seems not that straightforward the other way.

**Review by atmyers — APPROVED** [review]
